### ISSUE TITLE
[BUG] DP compress+DPLR+GPU results in bug in building neighbor list in LAMMPS

### ISSUE THREAD

**Issue by Liu-RX** [deepmodeling/deepmd-kit#3094]
### Bug summary

I was running DPMD of a system with 1023 water plus one OH- ion with DPLR model, when I found the bug. The DP energy model of the DPLR is compressed. I found for certain atomic geometries, the LAMMPS would fail on the neighbor-list-building step, with an error saying `Out of range atoms - cannot compute PPPM (src/KSPACE/pppm.cpp:1839)`. However, I do not know why these geometries could result in the bug, while others do not.
For example, the following geometry file would result in the bug above. 
[lammps-input.zip](https://github.com/deepmodeling/deepmd-kit/files/13777576/lammps-input.zip)

### DeePMD-kit Version

2.2.7

### TensorFlow Version

2.9.0

### How did you download the software?

Offline packages

### Input Files, Running Commands, Error Log, etc.

[lammps-input.zip](https://github.com/deepmodeling/deepmd-kit/files/13777737/lammps-input.zip)

### Steps to Reproduce

run lammps with the input files.

**Comment by wanghan-iapcm** [comment]
It seems that some atoms has large force predictions (result from ill-posed configuraiton) thus the simulation crashes. 
you may check the quality of the model by model deviation or reducing the dt to avoid ill-posed configurations.

**Comment by Liu-RX** [comment]
I reran the simulation with dt=0.05 fs, and it could successfully continue, but dt=0.1 fs does not work. The uncompressed model with dt = 0.5 fs works. The compressed model on the CPU also works. 
I did not train other models, so I cannot do the model deviation test. However, I have tested the model on some 64-water systems as a test set, and the atomic force deviation is ~0.8 eV/Angstrom. I thought the force deviation is already good enough.

**Comment by njzjz** [comment]
You didn't update the neighbor list every step, which is unsafe when an atom has high velocity.

**Comment by njzjz** [comment]
Do you mean you cannot get the same forces from uncompressed/compressed model on CPUs/GPUs?

**Comment by wanghan-iapcm** [comment]
> Do you mean you cannot get the same forces from uncompressed/compressed model on CPUs/GPUs?

This happens when distance between atoms falls out of the tabulation range, i.e. out of the range of training data.

**Comment by Liu-RX** [comment]
> You didn't update the neighbor list every step, which is unsafe when an atom has high velocity.

The simulation temperature is 330 K. I think the atomic velocities won't become too high.

**Comment by Liu-RX** [comment]
> Do you mean you cannot get the same forces from uncompressed/compressed model on CPUs/GPUs?

Sorry I did not explain the situation clearly enough. I just checked out some of the atomic forces from the compressed/uncompressed model on CPU/GPU.
I can get the same forces from the uncompressed model on CPU/GPU. I can also get the same forces from the compressed model on CPU/GPU. The atomic force difference on x/y/z on each atom between the compressed and uncompressed model is less than 0.01 eV/A.
The problem is that, if I set the number of neighbor list update steps to 5 (with `neigh_modify    every 5 delay 0 check no exclude type 1 3` command), the compressed model on GPU will fail on the 5th step. If I set the neighbor list update steps to 2, it will fail on the 2nd step. This failure does not come up for the compressed model on CPU. It also does not come up for the uncompressed model on either GPU or CPU. So I wonder if it is a bug.

**Comment by wanghan-iapcm** [comment]
> > You didn't update the neighbor list every step, which is unsafe when an atom has high velocity.
> 
> The simulation temperature is 330 K. I think the atomic velocities won't become too high.

depends on your training data. If they were generated at 100K, the 330K simulation would be definitely too high.  
it would be helpful if you report how did you generate your training data.

**Comment by Liu-RX** [comment]
The training data is generated by DP-GEN, and the MD simulation temperature in DP-GEN is also 330 K. The only difference is that the training data is 64-water system, while this DPMD system is a 1024-water system. I have tried the compressed model on 64-water and 256-water DPMD (at 330 K), and they both work well.

**Comment by njzjz** [comment]
Can you obtain the same energy/forces with different `every N` settings?

**Comment by Liu-RX** [comment]
Yes, I obtained the same results with different N. I just tried out N=2, 5, 8, and the energy/atomic forces are the same for both CPU and GPU.

**Comment by wanghan-iapcm** [comment]
> The training data is generated by DP-GEN, and the MD simulation temperature in DP-GEN is also 330 K. The only difference is that the training data is 64-water system, while this DPMD system is a 1024-water system. I have tried the compressed model on 64-water and 256-water DPMD (at 330 K), and they both work well.

In 1024 water system it seems that you have encountered local configurations that are never sample from the 64 water system. 
I would recommend doing dpgen in a boarder range of thermodynamic conditions to increase the data coverage. 

The reason why un-compressed dp model works is that it usually has the the ability to "extrapolate" to some extend, but it is not true for the compressed model.

**Comment by Liu-RX** [comment]
Thank you so much. Now I could understand the problem. I still have one more problem, though. Does 'a boarder range of thermodynamic conditions' mean sampling at higher temperatures or sampling at higher temperatures as well as in a wider density range?

**Comment by njzjz** [comment]
Could you share the model? I could take a look in this month.

**Comment by Liu-RX** [comment]
> Could you share the model? I could take a look in this month.

That will be most helpful! I will share it with you through Wechat after I confirm with my tutor.

**Comment by njzjz** [comment]
I got different results from uncompressed and compressed models and NaN forces on the GPU for the compressed model.

Uncompressed, CPU/GPU:
```
   Step         KinEng         PotEng         TotEng        Enthalpy         Temp          Press          Volume           Lx             Ly             Lz            Pxx            Pyy            Pzz      
         0   132.5647      -478777        -478644.44     -478675.52      334.16917     -1614.7996      30836.902      31.358618      31.358618      31.358618     -4095.0064     -228.18251     -521.20986    
         1   132.64479     -478777.08     -478644.44     -478680.93      334.37104     -1896.0611      30836.902      31.358618      31.358618      31.358618     -4053.6687     -419.20686     -1215.3077  
```

Compressed CPU:
```
   Step         KinEng         PotEng         TotEng        Enthalpy         Temp          Press          Volume           Lx             Ly             Lz            Pxx            Pyy            Pzz      
         0   132.5647      -478775.23     -478642.67     -478676.51      334.16917     -1758.4208      30836.902      31.358618      31.358618      31.358618     -4228.4375     -386.93289     -659.89187    
         1   132.64309     -478775.31     -478642.67     -478681.92      334.36676     -2039.5645      30836.902      31.358618      31.358618      31.358618     -4186.245      -578.45722     -1353.9913   
```

Compressed GPU:
```
   Step         KinEng         PotEng         TotEng        Enthalpy         Temp          Press          Volume           Lx             Ly             Lz            Pxx            Pyy            Pzz      
         0   132.5647      -478775.23     -478642.67     -478676.51      334.16917     -1758.4208      30836.902      31.358618      31.358618      31.358618     -4228.4375     -386.93289     -659.89187    
         1  -nan           -478775.31     -nan           -nan           -nan           -nan            30836.902      31.358618      31.358618      31.358618     -nan           -nan           -nan   
```

Compressed CPU:
```
ITEM: TIMESTEP
1
ITEM: NUMBER OF ATOMS
3070
ITEM: BOX BOUNDS xy xz yz pp pp pp
0.0000000000000000e+00 3.1358618000000000e+01 0.0000000000000000e+00
0.0000000000000000e+00 3.1358618000000000e+01 0.0000000000000000e+00
0.0000000000000000e+00 3.1358618000000000e+01 0.0000000000000000e+00
ITEM: ATOMS id type x y z vx vy vz fx fy fz
191 1 2.45916 3.72609 3.82534 -4.7833 4.26556 0.5949 -0.576842 1.00975 -0.262844
276 1 0.158369 1.78079 0.312709 0.102488 -0.65614 3.45001 -0.841808 0.508848 0.688868
550 1 0.13342 4.34131 0.380768 4.43509 2.98256 -4.41717 0.427465 -0.715532 -0.0771829
662 1 2.4422 0.118813 0.0923192 -0.923989 2.14932 -0.697069 -0.718864 -0.588667 -0.399676
1000 1 4.13363 3.20598 1.41086 -2.96468 -0.888981 0.237795 -1.84338 0.688436 1.06805
```

```
ITEM: TIMESTEP
1
ITEM: NUMBER OF ATOMS
3070
ITEM: BOX BOUNDS xy xz yz pp pp pp
0.0000000000000000e+00 3.1358618000000000e+01 0.0000000000000000e+00
0.0000000000000000e+00 3.1358618000000000e+01 0.0000000000000000e+00
0.0000000000000000e+00 3.1358618000000000e+01 0.0000000000000000e+00
ITEM: ATOMS id type x y z vx vy vz fx fy fz
191 1 2.45916 3.72609 3.82534 -nan -nan -nan -nan -nan -nan 
276 1 0.158369 1.78079 0.312709 -nan -nan -nan -nan -nan -nan 
550 1 0.13342 4.34131 0.380768 -nan -nan -nan -nan -nan -nan 
662 1 2.4422 0.118813 0.0923192 -nan -nan -nan -nan -nan -nan 
1000 1 4.13363 3.20598 1.41086 -nan -nan -nan -nan -nan -nan 
```

Although the model itself may have some problems, there might be a bug in the GPU version of `tabulate_fusion_se_a_grad_fifth_order_polynomial`, considering the potential energy is not NaN.

**Comment by wanghan-iapcm** [comment]
> Thank you so much. Now I could understand the problem. I further ran some 4096-water system DPMD, and found out that the uncompressed model could also give the same error. I still have one more problem, though. Does 'a boarder range of thermodynamic conditions' mean sampling at higher temperatures or sampling at higher temperatures as well as in a wider density range?

A broader range of both temperature and density is preferred.

**Comment by Liu-RX** [comment]
> > Thank you so much. Now I could understand the problem. I further ran some 4096-water system DPMD, and found out that the uncompressed model could also give the same error. I still have one more problem, though. Does 'a boarder range of thermodynamic conditions' mean sampling at higher temperatures or sampling at higher temperatures as well as in a wider density range?
> 
> A broader range of both temperature and density is preferred.

I see. I checked out again and confirmed that the training data is generated mostly at 330 K, some at 400 K. I will enlarge the temperature and the density range next time.

**Comment by njzjz** [comment]
I did some experiments on this issue, as shown below, but I don't know the problem.
(1) If the CPU is used, no NaN is produced;
(2) if the compressed model is used for `pair deepmd` and the uncompressed model is used for `fix dplr`, NaN can be produced;
(3) if the uncompressed mode is used for `pair deepmd`, and the compressed model is used for `fix dplr`, there is no NaN;
(4) if I remove `fix dplr`, there is no NaN.

So, NaN only happens with GPU + compressed `pair deepmd` + `fix dplr`, which really confuses me.

**Comment by Liu-RX** [comment]
> I did some experiments on this issue, as shown below, but I don't know the problem. (1) If the CPU is used, no NaN is produced; (2) if the compressed model is used for `pair deepmd` and the uncompressed model is used for `fix dplr`, NaN can be produced; (3) if the uncompressed mode is used for `pair deepmd`, and the compressed model is used for `fix dplr`, there is no NaN; (4) if I remove `fix dplr`, there is no NaN.
> 
> So, NaN only happens with GPU + compressed `pair deepmd` + `fix dplr`, which really confuses me.

This is also what I got when I came into this problem. Indeed it's puzzling. 😂

**Comment by njzjz** [comment]
Fixed by https://github.com/deepmodeling/deepmd-kit/pull/3157.
